### PR TITLE
SaferRemoveIndexConcurrently

### PR DESCRIPTION
**Pull request by marcelofern**
# Description

Adds a new Django migration operation class `SaferRemoveIndexConcurrently`. The name is self-explanatory, but here's a screenshot of the documentation (built with `make docs`) with more details (also available in the diff on the ante-penultimate commit).

image

### CODE REVIEW

File: src/django_pg_migration_tools/operations.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def safer_create_index(
         to_state: migrations.state.ProjectState,
         index: models.Index,
         model: type[models.Model],
-        operation: SaferAddIndexConcurrently,
+        operation: SaferAddIndexConcurrently | SaferRemoveIndexConcurrently,
```
Comment (delfick) [inline src/django_pg_migration_tools/operations.py:35]: can this instead be a Protocol class as the type you expect?

File: pyproject.toml [hunk 2]
```diff
@@ -21,7 +21,7 @@ where = ["src"]
 
 [project]
 name = "django_pg_migration_tools"
-version = "0.1.2"
+version = "0.1.3"
```
Comment (delfick) [inline pyproject.toml:24]: I'd encourage the previous commit updating dependencies and the part of this commit that changes the version number be a separate PR dedicated to releasing a new version rather than bundling it with this change.

File: src/django_pg_migration_tools/operations.py [hunk 3]
```diff
@@ -154,16 +116,116 @@ def _ensure_original_lock_timeout_is_reset(
             self.SET_LOCK_TIMEOUT_QUERY, {"lock_timeout": self.original_lock_timeout}
         )
 
-    # The following methods are necessary for Django to understand state
-    # changes.
-    def state_forwards(
-        self, app_label: str, state: migrations.state.ProjectState
+
+class SaferAddIndexConcurrently(
+    BaseIndexOperation, psql_operations.AddIndexConcurrently
```
Comment (delfick) [inline src/django_pg_migration_tools/operations.py:121]: is there anything stopping us from not using inheritence for composition? So that the SaferAddIndexConcurrently instance creates it's own instance of AddIndexConcurrently that it delegates to rather than being a subclass of AddIndexConcurrently?

File: src/django_pg_migration_tools/operations.py (hunk 1 above)
Comment (marcelofern) [reply src/django_pg_migration_tools/operations.py:35]: There should be the only two classes that create/drop indexes, so I'm not sure there're gains in using structural typing here if we're not aiming for flexibility. On the side, nominal typing provides stronger contracts too.

File: pyproject.toml (hunk 2 above)
Comment (marcelofern) [reply pyproject.toml:24]: Thanks. Are there any pros/cons to either approach I wouldn't be aware of? I thought that if a release is linked to a single change it would be best to keep it in the same PR so that it's easier to link to the scoped work from the tags  page. Similar to how some repos create a single release PR over multiple branches. Not particularly fussy about that either though.

File: src/django_pg_migration_tools/operations.py (hunk 3 above)
Comment (marcelofern) [reply src/django_pg_migration_tools/operations.py:121]: I'd surely prefer composition if the code action was "AddIndexConcurrently, please go do something for me".

But in this case the code action is more like "AddIndexConcurrently, I need your interface so that Django's migration apparatus can understand my as an index creation operation".

This includes having methods that will register application database state alterations for example. But more importantly, I'm interested in picking up changes from upstream on that specific interface.

For example, a recent change in Django (new in 5.1) has added the "category" attribute to operations. In this case, `category = OperationCategory.ADDITION` (or REMOVAL), wouldn't be picked up by our classes unless we inherit from upstream. Otherwise, we'd have to keep different interfaces for different Django versions, which is less desirable.

File: pyproject.toml (hunk 2 above)
Comment (delfick) [reply pyproject.toml:24]: yeah, certainly not a hard requirement, but my own sensibility is that releases should be separate as a general habit.

File: src/django_pg_migration_tools/operations.py (hunk 3 above)
Comment (delfick) [reply src/django_pg_migration_tools/operations.py:121]: fair enough, inheritence sounds reasonable then.

**Review by delfick — APPROVED** [review]

File: pyproject.toml (hunk 2 above)
Comment (marcelofern) [reply pyproject.toml]: [Here's](https://github.com/kraken-tech/django-pg-migration-tools/pull/19) a separated PR for the release. To be merged after this one is merged.

File: tests/django_pg_migration_tools/test_operations.py [hunk 4]
```diff
@@ -260,3 +260,170 @@ def test_when_not_allowed_to_migrate(self):
...
+
+        project_state = ProjectState()
+        project_state.add_model(ModelState.from_model(CharModel))
+        new_state = project_state.clone()
+
+        operation = operations.SaferRemoveIndexConcurrently(
+            "charmodel",
+            "char_field_idx",
+        )
+        # Proceed to try and add the index:
```
Comment (timb07) [inline tests/django_pg_migration_tools/test_operations.py:412]: copy/paste: "remove" instead of "add"

File: docs/usage/operations.rst [hunk 5]
```diff
@@ -109,3 +109,97 @@ Class Definitions
...
+      SET lock_timeout = 0;
+      -- Drop the index
+      DROP INDEX CONCURRENTLY IF EXISTS foo_idx;
+      -- Reset lock_timeout to its original value ("1s" as an example).
+      SET lock_timeout = '1s';
+
+    How to use
+    ----------
+
+    1. Remove the index to the relevant model:
```
Comment (timb07) [inline docs/usage/operations.rst:168]: copy/paste: should be "… from the relevant model"

**Review by timb07 — APPROVED** [review]
LGTM; just a couple of copy/paste typos to fix.

File: tests/django_pg_migration_tools/test_operations.py (hunk 4 above)
Comment (marcelofern) [reply tests/django_pg_migration_tools/test_operations.py]: Thx: 54d6cf9d4a37adffb3b0a4b70b7b301236f5cf93

File: docs/usage/operations.rst (hunk 5 above)
Comment (marcelofern) [reply docs/usage/operations.rst]: Thx: 50d63778177336e3f3eeba70599eecf3543a7359
